### PR TITLE
[msal-browser] Add custom storage option in configuration

### LINKED ISSUE #871: New Secure Option For Storage
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Browser:
- [ ] Chrome version XX
- [ ] Firefox version XX
- [ ] IE version XX
- [ ] Edge version XX
- [ ] Safari version XX

## Library version

<pre><code>
Library version: X.Y.Z

</code></pre>

## Current behavior
Msal js currently supports `localStorage` and `sessionStorage` for cache.  We would like to add other options, including a customer managed storage. This would let the app serialize their cache state and potentially share with other platforms

### LINKED ISSUE #1588: MSAL for VSCODE extension
Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [x] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description
The authentication using webView in VSCODE is really challenging becuase of it's design which doesn't allow the localStorage and sessionStorage instead it provides it's own custom storage options. It would be great to have this MSAL libraries to support the vscode kind for it's cache location and help developers life easy :-)

### LINKED ISSUE #2126: Web storage cache
## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description
Storing tokens in localStorage or sessionStorage is not really secured, because it is accessible from JS (XSS).

I don't think that sessionStorage is better than localStorage for apps that are fully protected by authentication, since the experience of re-login every time the user close the browser / tab is very bad and annoying.

What is the best approach for that? Is there any way to prevent MSAL.js from caching the JWT token in the web storage?

### PR DESCRIPTION
**Pull request by pkanher617**
This PR adds the ability for users to configure the storage location for data in the msal-browser package. 

The following changes are a part of this PR:
- An additional interface for browser storage APIs
- Updates to the configuration class to accept a custom implementation
- Refactoring changes to the BrowserCacheManager class
- Additional changes for refactoring
- Samples to show in-memory storage. Only works for popups.

### CODE REVIEW

File: lib/msal-browser/src/cache/IWindowStorage.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * Get all the keys from the window storage object as an iterable array of strings.
+     */
+    getWindowStorageKeys(): string[];
+
+    /**
+     * Returns true or false if the given key is present in the cache.
+     * @param key 
+     */
+    windowStorageContainsItem(key: string): boolean;
```
Comment (pkanher617) [inline lib/msal-browser/src/cache/IWindowStorage.ts:39]: should be containsKey, not item

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 2]
```diff
@@ -55,14 +55,11 @@ export class PublicClientApplication implements IPublicClientApplication {
     private readonly browserCrypto: CryptoOps;
 
     // Storage interface implementation
-    private readonly browserStorage: BrowserStorage;
+    private readonly browserStorage: BrowserCacheManager;
```
Comment (sameerag) [inline lib/msal-browser/src/app/PublicClientApplication.ts:58]: why use `storage` in the name if the base class name is changed anyway?

**Comment by MasashiSalvador57f** [comment]
Hi will this PR merged in the future? I found the github-actions said it would be stale & closed if w/o any activity, I need this feature in our product for storing idtoken in the secure way.

is there any help I can do? (like modifying some part of codes or so)

File: lib/msal-browser/src/cache/IWindowStorage.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Interface object for window storage. TODO: Add link to custom cache docs.
+ */
+export interface IWindowStorage {
+
+    /**
+     * Get the item from the window storage object matching the given key.
+     * @param key 
+     */
+    getWindowStorageItem(key: string): string;
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/IWindowStorage.ts:15]: I would prefer just `getItem, `getItem`, etc. This is arguably wordy and redundant.

**Comment by jasonnutter** [comment]
> Hi will this PR merged in the future? I found the github-actions said it would be stale & closed if w/o any activity, I need this feature in our product for storing idtoken in the secure way.
> 
> is there any help I can do? (like modifying some part of codes or so)

Yes, we will be merging this. It was blocked on other PRs, which are now merged.

File: lib/msal-browser/src/cache/BrowserStorage.ts [hunk 4]
```diff
@@ -2,40 +2,21 @@
...
-    constructor(clientId: string, cacheConfig: CacheOptions) {
-        super();
+    constructor(cacheLocation: string) {
         // Validate cache location
-        this.validateWindowStorage(cacheConfig.cacheLocation);
-
-        this.cacheConfig = cacheConfig;
-        this.windowStorage = window[this.cacheConfig.cacheLocation];
-        this.clientId = clientId;
+        this.validateWindowStorage(cacheLocation);
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserStorage.ts:17]: This function should catch errors that are thrown when acccessing local/session storage, and throw appropriate error about them not being supported/accessible in the browser.

Comment (jasonnutter) [reply lib/msal-browser/src/cache/BrowserStorage.ts:17]: That error should probably recommend using custom storage (or the built-in in memory option, if we decide to provide that).

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+import { Constants, PersistentCacheKeys, StringUtils, AuthorizationCodeRequest, ICrypto, CacheSchemaType, AccountEntity, IdTokenEntity, CredentialType, AccessTokenEntity, RefreshTokenEntity, AppMetadataEntity, CacheManager, CredentialEntity, ThrottlingEntity, ServerTelemetryCacheValue } from "@azure/msal-common";
+import { CacheOptions } from "../config/Configuration";
+import { BrowserAuthError } from "../error/BrowserAuthError";
+import { BrowserConfigurationAuthError } from "../error/BrowserConfigurationAuthError";
+import { BrowserConstants, TemporaryCacheKeys } from "../utils/BrowserConstants";
+import { BrowserStorage } from "./BrowserStorage";
+import { IWindowStorage } from "./IWindowStorage";
+
+// Cookie life calculation (hours * minutes * seconds * ms)
+const COOKIE_LIFE_MULTIPLIER = 24 * 60 * 60 * 1000;
```
Comment (jmckennon) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:14]: should this be in constants file?

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                this.browserStorage.setWindowStorageItem(key, stringVal);
+                if (this.cacheConfig.storeAuthStateInCookie) {
+                    this.setItemCookie(key, stringVal);
+                }
+                break;
+            }
+            case CacheSchemaType.THROTTLING:
+            case CacheSchemaType.TELEMETRY: {
+                this.browserStorage.setWindowStorageItem(key, JSON.stringify(value));
+                break;
```
Comment (jmckennon) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:117]: can these two cases be combined with the ACCOUNT, CREDENTIAL, and APP_METADATA cases? they have exactly the same logic, as far as i can see

Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:117]: I think this was potentially like this before this PR, but nonetheless

File: lib/msal-browser/src/cache/IWindowStorage.ts (hunk 3 above)
Comment (jmckennon) [reply lib/msal-browser/src/cache/IWindowStorage.ts:15]: +1

File: lib/msal-browser/src/utils/BrowserConstants.ts [hunk 7]
```diff
@@ -13,6 +13,8 @@ export const BrowserConstants = {
     CACHE_LOCATION_LOCAL: "localStorage",
     // Session storage constant string
     CACHE_LOCATION_SESSION: "sessionStorage",
+    // Custom storage constant string
+    CACHE_LOCATION_CUSTOM: "customStorage",
```
Comment (jmckennon) [inline lib/msal-browser/src/utils/BrowserConstants.ts:17]: are we expecting different kinds of storage to come out of this as well to have it be truly custom, or is the intent to just be an in-memory storage?

File: lib/msal-browser/src/cache/IWindowStorage.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Interface object for window storage. TODO: Add link to custom cache docs.
```
Comment (jmckennon) [inline lib/msal-browser/src/cache/IWindowStorage.ts:7]: reminder for this link + docs

Comment (pkanher617) [reply lib/msal-browser/src/cache/IWindowStorage.ts:7]: Will work on docs once we have agreed on the implementation. Thanks!

**Comment by asuka-sonoyama** [comment]
Hi,
Our team needs this functionality on production environment that will be go-live on next December.
Id just like to know when will you release this ?Do you have any plan or schedule of this?

**Comment by jasonnutter** [comment]
Closing, may revisit in the future.

**Comment by unmade-spangle** [comment]
Please revisit this and merge it, right now MSAL is COMPLETELY useless in browser extensions tyvm

**Pull request closed without merging**